// chess.ts/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chess_1 = require("./chess");
Object.defineProperty(exports, "Chess", { enumerable: true, get: function () { return chess_1.Chess; } });
var types_1 = require("./types");
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return types_1.State; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "WHITE", { enumerable: true, get: function () { return constants_1.WHITE; } });
Object.defineProperty(exports, "BLACK", { enumerable: true, get: function () { return constants_1.BLACK; } });
Object.defineProperty(exports, "PAWN", { enumerable: true, get: function () { return constants_1.PAWN; } });
Object.defineProperty(exports, "KNIGHT", { enumerable: true, get: function () { return constants_1.KNIGHT; } });
Object.defineProperty(exports, "BISHOP", { enumerable: true, get: function () { return constants_1.BISHOP; } });
Object.defineProperty(exports, "ROOK", { enumerable: true, get: function () { return constants_1.ROOK; } });
Object.defineProperty(exports, "QUEEN", { enumerable: true, get: function () { return constants_1.QUEEN; } });
Object.defineProperty(exports, "KING", { enumerable: true, get: function () { return constants_1.KING; } });
Object.defineProperty(exports, "EMPTY", { enumerable: true, get: function () { return constants_1.EMPTY; } });
Object.defineProperty(exports, "SQUARES", { enumerable: true, get: function () { return constants_1.SQUARES; } });
Object.defineProperty(exports, "FLAGS", { enumerable: true, get: function () { return constants_1.FLAGS; } });
Object.defineProperty(exports, "BITS", { enumerable: true, get: function () { return constants_1.BITS; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "rank", { enumerable: true, get: function () { return utils_1.rank; } });
Object.defineProperty(exports, "file", { enumerable: true, get: function () { return utils_1.file; } });
Object.defineProperty(exports, "algebraic", { enumerable: true, get: function () { return utils_1.algebraic; } });
